src/System.Linq.Expressions/src/System/Dynamic/Utils/TypeExtensions.cs

```diff
@@ -44,6 +37,11 @@ private static bool MatchesArgumentTypes(this MethodInfo mi, Type[] argTypes)
             Debug.Assert(mi != null);
```

Comment:
In this case I just noticed it going through the code.

For the more general answer, let's take the Innerloop PortableLinux Release x64 Build and Test case as an example. Following the link to https://ci.dot.net/job/dotnet_corefx/job/master/job/portablelinux_release_prtest/3664/ and following the two links listed as "Indication 1" and "Indication 2" we get taken to different points in the logs. These vary in how useful they are, but the second one is pretty clear (and the first a knock-on from it), the test failed because "The configuration file has been changed by another program."

That case sounds like it has nothing to do with your change and is a problem elsewhere. These happen due to bad non-deterministic tests, an issue with the CI, a good non-deterministic test sometimes finding a real problem in the codebase, etc. If you can search for related issues already open and either add a note to it so people can use the data or open a new issue (maybe not if it's related to networking, as networking can always have networking issues). This seems to be related to #18431 so I added a note there.

But of course it can be a problem with your PR, which is the whole problem. Most of the rest here are that, but unfortunately `Debug.Assert` does not fail in the nicest way. Still, if we take Innerloop CentOS7.1 Debug x64 Build and Test as an example and go to https://ci.dot.net/job/dotnet_corefx/job/master/job/centos7.1_debug_prtest/6388/ then following the link to the console output we see `System.Diagnostics.Debug+DebugAssertException` within the stack trace. Because of the way `DebugAssertException` takes things down it can be misleading in terms of what test seemed to trigger it (it's reported as having failed in a S.L.Queryable test) but looking at the stack trace starting from that mention of `DebugAssertException` shows that just before `FailFastWithException` (which debug assertions use to bail out) `System.Linq.Expressions.dll!System.Dynamic.Utils.TypeExtensions.MatchesArgumentTypes(System.Reflection.MethodInfo, System.Type[])` so that's the first place to look for where the assertion failed.

**Why** the assertion failed can be trickier to figure out, but in this case that it would have to be removed is mentioned in one of the comments above :)